taxonomy/providers/xblock_metadata.py

```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
+"""
+Abstract base class for xblock metadata providers.
+
+All host platform must implement this provider in order for taxonomy to work.
+"""
+from abc import abstractmethod
+
+
+class XBlockMetadataProvider:
+    """
+    Abstract base class for xblock metadata providers.
+
+    All abstract methods must be implemented for taxonomy's normal functionality.
+    """
+
+    @abstractmethod
+    def get_xblocks(self, xblock_ids):
+        """
+        Get a list of xblocks matching the xblock ids provided in the argument.
+        Include content of all children xblocks in content
+
+        Arguments:
+          xblock_ids(list<str>): A list of UUIDs in the form of a string.
+
+        Returns:
+          list<dict>: A list of xblocks in the form of dictionary.
+            Dictionary object must have the following keys
+            1. key: xblock usage key
+            2. content_type: xblock content type
+            3. content: xblock text content
+        """
+
+    @abstractmethod
+    def get_all_xblocks(self):
```

Comment:
How many XBlocks would this return? Wouldn't the number of XBlocks on a site like edx.org be way too high to have any "get all" method that works?

Reply:
@bradenmacdonald True, it is not possible. I am still not sure how we can run the process for all xblocks the first time.

We can probably change this method to return all units and videos in a course.
Then create a custom django command which uses https://github.com/open-craft/taxonomy-connector/blob/d1c337a721fd857942e5a5d31893cd4e8b4f2f9c/taxonomy/providers/course_metadata.py#L36-L48 and this method to start processing all units/videos xblocks one by one.

Something similar to [refresh_course_skills](https://github.com/open-craft/taxonomy-connector/blob/master/taxonomy/management/commands/refresh_course_skills.py) command.

Reply:
That sounds like a good approach yes. In general it should be some manual/office/special process that you can trigger which runs course by course to do it; that way it's also possible to pause and resume if there's an issue, or to only focus on current courses and ignore older ones, if people want to.